Nhaimerl voting tutorial (#2758)
dfinity/portal: docs/developer-docs/defi/rosetta/icp_rosetta/construction_api/voting/follow.mdx

---
keywords: [intermediate, rosetta, tutorial, spawn, neuron, staking, voting, proposal, follow]
---

import { MarkdownChipRow } from "/src/components/Chip/MarkdownChipRow";

# Voting on proposals

<MarkdownChipRow labels={["Intermediate", "Tutorial" ]} />

## Overview
This section will show you how to use Rosetta to follow existing neurons. 
If you do not want to have to vote on every proposal yourself, you can decide to follow other neurons. This essentially means that you are going to vote the same way on a given proposal as the neuron that you are following. 

The `FOLLOW` operation sets a follow rule for a neuron.

> The Governance canister smart contract will automatically deduce the vote of the following neurons from the votes of the followees during the voting.

Reviewer: ```suggestion
The governance canister smart contract will automatically deduce the vote of the following neurons from the votes of the followees during the voting.
```

<div class="formalpara-title">

**Calling `FOLLOW` with a hotkey:**

</div>

```json
curl --location '0.0.0.0:8081/construction/payloads'  --header 'Content-Type: application/json' --data '{
  "network_identifier": {
    "blockchain": "Internet Computer",
    "network": "00000000000000020101"
  },
  "public_keys": [
    {
      "hex_bytes": "047a83e378053f87b49aeae53b3ed274c8b2ffbe59d9a51e3c4d850ca8ac1684f7131b778317c0db04de661c7d08321d60c0507868af41fe3150d21b3c6c757367",
      "curve_type": "secp256k1"
    }
  ],
  "operations": [
    {
      "operation_identifier": {
        "index": 0
      },
      "type": "FOLLOW",
      "account": {
        "address": "8b84c3a3529d02a9decb5b1a27e7c8d886e17e07ea0a538269697ef09c2a27b4"
      },
      "metadata": {
        "neuron_index": 1,
        "followees": [
          3094748712371737240
        ],
        "topic": 0,
        "controller": {
          "public_key": {
            "hex_bytes": "047a83e378053f87b49aeae53b3ed274c8b2ffbe59d9a51e3c4d850ca8ac1684f7131b778317c0db04de661c7d08321d60c0507868af41fe3150d21b3c6c757367",
            "curve_type": "secp256k1"
          }
        }
      }
    }
  ],
  "metadata": null
}'
```

:::info

> The `followees` metadata field contains list of unique neuron identifiers assigned by the Governance canister smart contract, not the list of neuron indices chosen by the caller.

Reviewer: ```suggestion
The `followees` metadata field contains list of unique neuron identifiers assigned by the governance canister smart contract, not the list of neuron indices chosen by the caller.
```

> You can obtain unique neuron identifiers of you your neurons from the `neuron_id` metadata field of the `STAKE` and `NEURON_INFO` operations.

Reviewer: ```suggestion
You can obtain unique neuron identifiers of your neurons from the `neuron_id` metadata field of the `STAKE` and `NEURON_INFO` operations.
```

:::

You can see the followee being added to the list of followees when fetching the `NeuronInfo`. 

> An example implementation in python looks like the following.

Reviewer: ```suggestion
An example implementation in Python resembles the following:
```
